### ISSUE TITLE
z-way GPIO daughterboard

### ISSUE THREAD

**Issue by snizzleorg** [openhab/openhab-docker#3]
how does one use the GPIO mounted z-way wave daughterboard with this?

This works normally like a serial port /dev/ttyAMA0

with openhab 1.8 one has to add 

```
 -Dgnu.io.rxtx.SerialPorts=/dev/ttyAMA0 \

```

and 
```
zwave:port=/dev/ttyAMA0
```
in the openhab.cfg

to the start script. Any way to mount /dev/ttyAMA0 to something the dockerized openhab will use?

**Comment by cyberkov** [comment]
Hello! 
Yes you can use the ``--device /dev/ttyAMA0`` to mount the device into the container.  In addition you can use the ``EXTRA_JAVA_OPTS`` Environment variable to set the flags for java. Please be aware that this is a current version of openhab2. Zwave will be configured in /things or through the paperui. 
Cheers Hannes

**Comment by snizzleorg** [comment]
@cyberkov Thanks!

so a docker-compose.yml like 
```
---
openhab:
  image: 'openhab/openhab:armhf-online'
  restart: always
  ports:
    - "8080:8080"
    - "8443:8443"
    - "5555:5555"
  net: "host"
  volumes:
    - '/etc/localtime:/etc/localtime:ro'
    - '/etc/timezone:/etc/timezone:ro'
    - '/opt/openhab/userdata:/openhab/userdata'
    - '/opt/openhab/conf:/openhab/conf'
  devices:
     - "/dev/ttyAMA0:/dev/ttyUSB0"
  environment:
     - "EXTRA_JAVA_OPTS":"-Dgnu.io.rxtx.SerialPorts=/dev/ttyAMA0"
  command: "server"
```
should do the trick?
I am aware that this is openhab2 - this is why I wanted to use a docker container. any suggests on how to best get the openhab1 config "imported"?

**Comment by cyberkov** [comment]
Hi! 
Remove the ``:/dev/ttyUSB0`` part and it should work :-)
I'm on my mobile right now but you can check my my-openhab-config repo to see a working oh2 config. 
Cheers Hannes

**Comment by snizzleorg** [comment]
its working.
```
---
openhab:
  image: 'openhab/openhab:armhf-online'
  restart: always
  ports:
    - "8080:8080"
    - "8443:8443"
    - "5555:5555"
  net: "host"
  volumes:
    - '/etc/localtime:/etc/localtime:ro'
    - '/etc/timezone:/etc/timezone:ro'
    - '/opt/openhab/userdata:/openhab/userdata'
    - '/opt/openhab/conf:/openhab/conf'
  devices:
     - "/dev/ttyAMA0"
  environment:
     - "EXTRA_JAVA_OPTS:-Dgnu.io.rxtx.SerialPorts=/dev/ttyAMA0"
  command: "server"

```

I will have to check if the z-wave stuff really is able to communicate when I'm home. But so far I can at least see the configuration options for the z-wave module...

**Comment by snizzleorg** [comment]
hm... unfortunately I get 
```
2016-04-06 13:06:54.549 [INFO ] [me.event.ThingStatusInfoChangedEvent] - 'zwave:serial_zstick:a9229ce2' changed from ONLINE to OFFLINE (COMMUNICATION_ERROR): Serial Error: Port /dev/ttyAMA0 does not exist
```

**Comment by snizzleorg** [comment]
the device is actually available but I guess maybe the permissions are wrong?
```
ls -l /dev/
total 0
lrwxrwxrwx 1 root root          13 Apr  6 13:10 fd -> /proc/self/fd
crw-rw-rw- 1 root root      1,   7 Apr  6 13:10 full
c--------- 1 root root     10, 229 Apr  6 13:10 fuse
drwxrwxrwt 2 root root          40 Jan  1  1970 mqueue
crw-rw-rw- 1 root root      1,   3 Apr  6 13:10 null
lrwxrwxrwx 1 root root           8 Apr  6 13:10 ptmx -> pts/ptmx
drwxr-xr-x 2 root root           0 Apr  6 13:10 pts
crw-rw-rw- 1 root root      1,   8 Apr  6 13:10 random
drwxrwxrwt 2 root root          40 Apr  6 13:10 shm
lrwxrwxrwx 1 root root          15 Apr  6 13:10 stderr -> /proc/self/fd/2
lrwxrwxrwx 1 root root          15 Apr  6 13:10 stdin -> /proc/self/fd/0
lrwxrwxrwx 1 root root          15 Apr  6 13:10 stdout -> /proc/self/fd/1
crw-rw-rw- 1 root root      5,   0 Apr  6 13:10 tty
crw-rw---- 1 root dialout 204,  64 Apr  6 13:10 ttyAMA0
crw-rw-rw- 1 root root      1,   9 Apr  6 13:10 urandom
crw-rw-rw- 1 root root      1,   5 Apr  6 13:10 zero
```

**Comment by cyberkov** [comment]
hmm. the user openhab (with which the java process runs) is in the group dialout. So it should work...

**Comment by snizzleorg** [comment]
yep... thats what i checked as well. any chance that the "EXTRA_JAVA_OPTS:-Dgnu.io.rxtx.SerialPorts=/dev/ttyAMA0" is somehow to blame?

I don't see the parameter in the with ps -ax
```
ps -ax | grep java
10295 ?        Sl     2:51 /usr/lib/jvm/java-8-oracle/bin/java -Dopenhab.home=/openhab -Dopenhab.conf=/openhab/conf -Dopenhab.runtime=/openhab/runtime -Dopenhab.userdata=/openhab/userdata -Dorg.osgi.service.http.port=8080 -Dorg.osgi.service.http.port.secure=8443 -Xmx256m -Djava.awt.headless=true -Djava.endorsed.dirs=/usr/lib/jvm/java-8-oracle/jre/lib/endorsed:/usr/lib/jvm/java-8-oracle/lib/endorsed:/openhab/runtime/karaf/lib/endorsed -Djava.ext.dirs=/usr/lib/jvm/java-8-oracle/jre/lib/ext:/usr/lib/jvm/java-8-oracle/lib/ext:/openhab/runtime/karaf/lib/ext -Dkaraf.instances=/openhab/runtime/karaf/instances -Dkaraf.home=/openhab/runtime/karaf -Dkaraf.base=/openhab/userdata -Dkaraf.data=/openhab/userdata -Dkaraf.etc=/openhab/runtime/karaf/etc -Djava.io.tmpdir=/openhab/userdata/tmp -Djava.util.logging.config.file=/openhab/userdata/etc/java.util.logging.properties -Dkaraf.startLocalConsole=true -Dkaraf.startRemoteShell=true -classpath /openhab/runtime/karaf/lib/boot/org.apache.karaf.diagnostic.boot-4.0.3.jar:/openhab/runtime/karaf/lib/boot/org.apache.karaf.jaas.boot-4.0.3.jar:/openhab/runtime/karaf/lib/boot/org.apache.karaf.main-4.0.3.jar:/openhab/runtime/karaf/lib/boot/org.osgi.core-6.0.0.jar org.apache.karaf.main.Main
```

**Comment by cyberkov** [comment]
According to [this thread](https://community.openhab.org/t/solved-how-to-define-a-serial-port-on-raspberry-for-openhab2-for-my-razberry-zwave-binding/6693/8) this should work. Unfortunately I haven't tried it myself yet, sorry. But I am very keen to get it it running as I expect this will come up more often with docker :)

**Comment by snizzleorg** [comment]
ok. I see but just to be clear I use docker-compose and have the environment part in the docker-compose.yml.

I also tried to export the env variable via: 

```
 export EXTRA_JAVA_OPTS="-Dgnu.io.rxtx.SerialPorts=/dev/ttyAMA0"
```

and then running docker-compose

**Comment by cyberkov** [comment]
Ah! now I see it:
```
 environment:
     - "EXTRA_JAVA_OPTS:-Dgnu.io.rxtx.SerialPorts=/dev/ttyAMA0"
```
should be
```
 environment:
     - "EXTRA_JAVA_OPTS=-Dgnu.io.rxtx.SerialPorts=/dev/ttyAMA0"
```
it should be an = sign.
exporting Variables in the host does not copy them to the container, you need to specify them in compose :)

**Comment by snizzleorg** [comment]
cool!
its working.

Maybe you can put this working docker-compose into the read-me
```
---
openhab:
  image: 'openhab/openhab:armhf-online'
  restart: always
  ports:
    - "8080:8080"
    - "8443:8443"
    - "5555:5555"
  net: "host"
  volumes:
    - '/etc/localtime:/etc/localtime:ro'
    - '/etc/timezone:/etc/timezone:ro'
    - '/opt/openhab/userdata:/openhab/userdata'
    - '/opt/openhab/conf:/openhab/conf'
  devices:
     - "/dev/ttyAMA0:/dev/ttyAMA0"
  environment:
     - "EXTRA_JAVA_OPTS=-Dgnu.io.rxtx.SerialPorts=/dev/ttyAMA0"
  command: "server"

```

**Comment by cyberkov** [comment]
Thank you for testing it! I will add it to the Readme.

**Comment by snizzleorg** [comment]
I added a directory for the add ons, this is my working version for the rpi2:

```
---
openhab:
  image: 'openhab/openhab:armhf-online'
  restart: always
  ports:
    - "8080:8080"
    - "8443:8443"
    - "5555:5555"
  net: "host"
  volumes:
    - '/etc/localtime:/etc/localtime:ro'
    - '/etc/timezone:/etc/timezone:ro'
    - '/opt/openhab/userdata:/openhab/userdata'
    - '/opt/openhab/conf:/openhab/conf'
    - '/opt/openhab/addons:/openhab/addons'
  devices:
     - "/dev/ttyAMA0:/dev/ttyAMA0"
  environment:
     - "EXTRA_JAVA_OPTS=-Dgnu.io.rxtx.SerialPorts=/dev/ttyAMA0"
  command: "server"
```

**Comment by cyberkov** [comment]
Of course :-) RPi3 is the first 64bit RPi. Unfortunately Dockerhub does not support multiple tags per image yet. I think this would make it more clear which one to use.

**Comment by snizzleorg** [comment]
ok. my bad. I was somehow under the impression the rpi2 already was 64bit...

**Comment by umiddelb** [comment]
> Of course :-) RPi3 is the first 64bit RPi.

It is the first RPi with a 64bit SoC, but currently still uses a 32 bit kernel and userland ...

**Comment by polarathene** [comment]
I've tried to follow what was here. My system(Arch Linux, x86_64) is using SystemD and `/etc/timezone` directory exists but is empty, it seemed to cause a problem, I'm not sure what the equivalent is. Omitting that line let me start the container, I've heard that Docker might use the host systems timezone or a environment var could set it(new to Docker and OpenHAB).

Other issue I've run into is getting the z-wave binding working. I'm using a AeonLabs USB Z-Stick, I've forwarded it with my compose file and included the Java environment var like shown here, but still no read access.

`ls -l /dev` shows:
Host Machine: `crw-rw----  1 root uucp      166,   0 Jun 13 14:05 ttyACM0`
Container: `crw-rw---- 1 root   14 166,   0 Jun 13 16:46 ttyACM0`
The container has a user but no group like all the other listings, I don't think the missing third field(everyone) for permissions would be an issue, but perhaps the lack of an assigned group is? I guess it may also be due to [note 4](https://github.com/openhab/openhab/wiki/Serial-Binding#note4) mentioned on the wiki for Serial Binding, it refers to openhab being a member of the dialout group, when using Docker would that instead mean giving Docker access to the group? As can be seen in my `ls -l /dev` above, I have `uucp` on the host machine as group instead of `dialout`, I assume I'd treat it the same as `dialout`?

My `docker-compose.yml`:

```yml
---
openhab:
  image: 'openhab/openhab:amd64-online'
  restart: always
  ports:
    - "8080:8080"
    - "8443:8443"
    - "5555:5555"
  net: "host"
  volumes:
    - '/etc/localtime:/etc/localtime:ro'
    - '~/openhab/userdata:/openhab/userdata'
    - '~/openhab/conf:/openhab/conf'
    - '~/openhab/addons:/openhab/addons'
  devices:
    - '/dev/ttyACM0'
  environment:
    - "EXTRA_JAVA_OPTS=-Dgnu.io.rxtx.SerialPorts=/dev/ttyACM0"
  command: "server"
```

I map volumes to my home directory as previously I tried kitematic, though it did not seem to support supplying my own compose yml or adding devices. Although it was suggested earlier to remove the mapping for devices, I don't think that would be an issue? It also appears the environment var for java may not always be needed, I'm presently using a desktop PC with it the controller plugged in via USB rather than a daughterboard on a Pi.

**Comment by polarathene** [comment]
I've been able to verify connecting to the zwave controller via another docker container(using nodejs app and openzwave instead of openhab). I've skipped compose and just gone with the example docker run command with changes above. It works when I add `--user="root"`, the extra java options are not required(I have tried them without the --user flag and had no luck). So this has something to do with running as the `openhab` default user.

Any suggestions on how to debug?

**Comment by polarathene** [comment]
**SOLVED!**

It appears to be due to the group `uucp`. On Arch Linux(my host) the gid is `14`, however on the openHAB Ubuntu container the gid for uucp is `10`. This results in the group being unknown but I guess the problem is due to the `openhab` user not having this gid in their group rather than it being listed in `/etc/groups`.

I ran `find / -gid 10` inside the container, no results returned. I assume this gid will depend on host OS(no idea if the controller device is treated as uucp or dialout on Ubuntu.), would it be safe to also add gid of `14` to the `openhab` user? Perhaps assign a group name of `uucp-alt` or `uucp-gid14` for clarity. If there is no confirmation of `uucp` being used by any users with the gid of `10` then, perhaps `14` can be the default gid for `uucp`?

Here is what I'm doing now:

start.sh:
```
# build docker image with support for uucp using gid 14
docker build -t 'local-openhab:dockerfile' .

# force removes active containers (probably not a great idea)
docker rm -f `docker ps --no-trunc -aq`

# run the docker image we built
docker run \
        --name openhab \
        --net=host \
        --device=/dev/ttyACM0 \
        -v /etc/localtime:/etc/localtime:ro \
        -v ~/openhab/conf:/openhab/conf \
        -v ~/openhab/userdata:/openhab/userdata \
        -d \
        --restart=always \
        local-openhab:dockerfile
```
Dockerfile:
```
FROM openhab/openhab:amd64-online
USER root
RUN groupadd --gid 14 uucp-gid14
RUN usermod -aG uucp-gid14 openhab
USER openhab
```

**Comment by cyberkov** [comment]
Hello!

Sorry for my late response. I'm glad you found the solution to this issue :)
I am not sure if adding more gid's than dialout (which is ubuntus default) to the user would solve the problem properly for all users as I think the permissions on the device should be handled by udev of the docker host. If you could set the owner with a udev rule ( [arch wiki udev](https://wiki.archlinux.org/index.php/Udev#Setting_static_device_names) ) i guess it will work without modifications :)

Another option would be to enhance the entrypoint.sh with a directory for shellscripts (something like [run-parts](http://manpages.ubuntu.com/manpages/precise/en/man8/run-parts.8.html)) to give the user the ability to modify the behavior as needed (e.g. install software like pip etc).

Cheers Hannes

**Comment by polarathene** [comment]
@cyberkov from what I've read around online, gid 10 is usually the `wheel` group on many linux distro's not `uucp` like on Ubuntu. `uucp` is related to serial ports as far as I understand, I came across it when working with Arduino as well.

So your solution is to adjust the gid on the host to that of Ubuntu's `dialout` group(gid 20) via udev?

I've done some research and most commonly `dialout` has a gid of `20` and `uucp` a gid of `14`.
- Debian and deriviatives such as Ubuntu - dialout:20, uucp:10 - [ref](http://www.togaware.com/linux/survivor/Standard_Groups.html)
- Arch - no dialout, uucp:14 - [ref](https://wiki.archlinux.org/index.php/DeveloperWiki:UID_/_GID_Database)
- Alpine - dialout:20, uucp:14 
- RedHat: no dialout, uucp:14 [ref](https://www-uxsup.csx.cam.ac.uk/pub/doc/redhat/redhat7/rhl-rg-en-7.0/s1-sysadmin-usr-grps.html)
- Linux From Scratch: dialout:10, uucp:32 [ref](http://www.linuxfromscratch.org/clfs/view/clfs-2.0/arm/final-preps/creatingfiles.html)
- openSUSE: dialout:16, uucp:14
- Fedora: dialout:18, no uucp . Past versions had no dialout and uucp:14 [ref](https://docs.fedoraproject.org/en-US/Fedora/12/html/Deployment_Guide/s1-users-groups-standard-groups.html)

Distro's without ref link, information was sourced via `getent group dialout` and `getent group uucp` via Virtual Machines.

[Debian Wiki](https://wiki.debian.org/SystemGroups):
- dialout: Full and direct access to serial ports. Members of this group can reconfigure the modem, dial anywhere, etc. 
- uucp: The uucp user and group is used by the UUCP subsystem. It owns spool and configuration files. Users in the uucp group may run uucico. 

[Original use of `uucp` is deprecated, now used for serial ports?](http://docstore.mik.ua/orelly/other/Docs/oreilly/other2/puis3rd/0596003234_puis3-chp-5-sect-1.html)
- `Controls ownership of the Unix serial ports. (uucp traditionally managed the UUCP system, which is now deprecated.)`

> UUCP was a collection of utilities to support "Unix to Unix copy" tools before there was an internet. HOW it dialed out was up to the setup script, and could use any method. - [source](http://forums.fedoraforum.org/showthread.php?t=309763)

[This](http://unix.stackexchange.com/questions/152326/why-does-uucp-take-group-10) was a common rank 1 result in my searches.
- Link to Ubuntu docs is broken and should be [this](http://manpages.ubuntu.com/manpages/precise/man1/uucp.1posix.html), it documents the legacy use of `uucp`.

[This](https://fedorahosted.org/fldigi/wiki/Documentation/HOWTO/Serial_Port_Setup) link describes how serial port access group name of `uucp` or `dialout` depends on distro.
- Another confirmation [here](http://wiki.mattercontrol.com/Running_on_Linux), they advise to keep the gid/group the same as what the OS uses for other serial devices.
- Ubuntu [treating the same `ttyACM0` address/device as `dialout` group](http://askubuntu.com/questions/58119/changing-permissions-on-serial-port) vs `uucp` group on Arch.

---

Reasons to include `uucp`:
- `uucp` is used by some distro's instead of `dialout` for serial devices
- gid of 14 is fairly standard for `uucp` and perhaps a more consistent gid than those for `dialout`
- `uucp` group with gid 10 in the Docker image is unlikely to be used/accessed for serial devices, nor does openHAB have a use for the legacy `uucp` usage. Changing the gid to 14 should have no negative impact but result in supporting more host OS without the need for asking users to tweak the gid their device mounts as(may not always be udev). This aids in portability of the container.
- If you switch to Alpine for a smaller image at a later date, which is popular for Docker images to base off, this will not affect users as the groups `dialout:20` and `uucp:14` would be the same allowing for seamless transition.

Regardless, since `dialout` can vary with it's gid across various host OS, you'll probably want to add to wiki/docs about this issue popping up for users not using the same host OS as what the Docker container for openHAB uses. Then either a way to pass the gids for openHAB to use or udev rule advice.

**Comment by cyberkov** [comment]
Hello!

I think this issue should be solved with #22 
Closing it.

Cheers
Hannes
